### PR TITLE
Require all Python to pass mypy unless specifically excluded

### PR DESCRIPTION
**Pull request by callahad**
This is a maximal response to #11271: instead of specifying files to _include_ in mypy checking, all files are considered unless explicitly _excluded._

We may want to just take the first commit instead, and focus only on the `synapse/` directory. Or some middle ground.

For example: instead of getting this whole repo to pass Mypy in a single run, we may want to:

- Explicitly exempt everything in `contrib/`, `tests/`, etc.
- Check standalone scripts, schema migrations, etc. separately and individually, instead of as part of a single omnibus mypy run for the `synapse` module (doing so separately is [encouraged](https://github.com/python/mypy/issues/4008#issuecomment-332032784) by mypy, but the `scripts_are_modules` setting kind of works around it).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Review by DMRobertson — APPROVED** [review]
LGTM. We might want to ensure that the Python outside of the `synapse` package is linted too? But that's a separate change.

**Comment by callahad** [comment]
Would appreciate a second review before merging, since this does pretty significantly change our static type checking strategy across the repo.

**Review by reivilibre — APPROVED** [review]
I also think this is sensible and preferable — good to know that you can get the exclude functionality with a monster regex.

File: mypy.ini
Comment (richvdh) [inline mypy.ini:19]: won't this iterate into virtualenvs etc?

File: mypy.ini
Comment (richvdh) [inline mypy.ini:43]: I guess I'm not *entirely* convinced we need to enforce mypy for `contrib` stuff, but :man_shrugging:

File: mypy.ini
Comment (richvdh) [inline mypy.ini:35]: since this is enabling comments... maybe we could add a comment to say what it is doing?

```suggestion
exclude = (?x)   # enable verbose REs, see https://docs.python.org/3/library/re.html#re.X
```

File: mypy.ini
Comment (callahad) [reply mypy.ini:19]: Mypy won't recurse into hidden directories, so as long as your venv is in `.venv/` you're golden.

Are there any other common paths we should specifically exclude?

File: mypy.ini
Comment (callahad) [reply mypy.ini:43]: Yeah, me either. Heck, I'm not even sure we should be shipping `contrib/` in this repo at all 😉 

I'll switch this to a blanket exclusion of `|contrib/.*`

File: mypy.ini
Comment (DMRobertson) [reply mypy.ini:19]: The README (and maybe docs) recommends putting the venv in an `env` directory with `python3 -m venv ./env`. I did this on day one.

Comment (callahad) [reply mypy.ini:19]: Hm. Y'all should definitely switch to `.venv`, but I'll add an exclusion for `env/` for now ;)

Comment (DMRobertson) [reply mypy.ini:19]: Happy to; I was just getting at that we ought to update the README! (and perhaps for sydent/sygnal too?)

Comment (richvdh) [reply mypy.ini:19]: I'm not enthusiastic about adding this sort of constraint to people's dev environments. I'm not aware of any rule that says "thou shalt have no virtualenvs in your working tree save that they be called `.venv`", and I don't necessarily expect anyone else to be aware of such a rule.

Please can we just list the directories we care about?

Comment (richvdh) [reply mypy.ini:19]: To add to this: my main concern is if every different project starts having its own, specific rules about where the venv should go. I really don't want to have to start remembering different places for different projects.

So, if this is a thing we want to do, we have to get into a big debate about what The One True Place For Your Virtualenv should be, which isn't a fight I particularly want to have.

Comment (richvdh) [reply mypy.ini:19]: also also: I have several *different* venvs in my working copy, with different selections of libraries, and python interpreters, so in any case just excluding `env` doesn't help a great deal.

File: mypy.ini
Comment (richvdh) [inline mypy.ini:19]: do we need to include `.ci/scripts` here?

Comment (callahad) [reply mypy.ini:19]: Ah, yep!

File: mypy.ini
Comment (callahad) [reply mypy.ini:19]: /me gestures wildly at "I'll add an exclusion for env/ for now ;)" and 515ca6bb6c085cd94d1ccf11bd07faf81ea95fde from an hour ago

Comment (richvdh) [reply mypy.ini:19]: I'm yet to be convinced, I'm afraid.

Comment (richvdh) [reply mypy.ini:19]: basically anything that's not checked in - my whole point is that this shouldn't break when people have things in their working copy that you don't expect.

Comment (richvdh) [reply mypy.ini:19]: I guess maybe it would be ok to ignore things that are in the user's various gitignore configs, but that feels a bit magical.

Comment (callahad) [reply mypy.ini:19]: In [Apache voting terms](http://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions), how strongly are your feelings on this issue?

Assume the alternative would be to only cover `synapse/` and `tests/` with mypy, allowing other files in the repo to go without type checking.

Comment (richvdh) [reply mypy.ini:19]: > In [Apache voting terms](http://www.apache.org/foundation/voting.html#expressing-votes-1-0-1-and-fractions), how strongly are your feelings on this issue?

-0.9ish.

I like this way of resolving the dispute.

> Assume the alternative would be to only cover synapse/ and tests/ with mypy, allowing other files in the repo to go without type checking.

Well, we could surely add other directories - `scripts-dev`, `scripts`, etc. As noted elsewhere, we already have to have multiple entries in the list.

File: mypy.ini
Comment (richvdh) [inline mypy.ini:21]: `synctl` is missing from this list.

**Pull request closed without merging**
